### PR TITLE
Start etcd in integration tests with TestMain

### PR DESCRIPTION
**Pull request by mikedanese**
https://github.com/kubernetes/kubernetes/issues/47089

### CODE REVIEW

File: test/integration/framework/BUILD [hunk 1]
```diff
@@ -10,10 +10,14 @@ load(
 go_library(
     name = "go_default_library",
     srcs = [
+        "etcd.go",
         "master_utils.go",
         "perf_utils.go",
         "serializer.go",
     ],
+    data = [
+        "@com_coreos_etcd//:etcd",
```
Comment (spxtr) [inline test/integration/framework/BUILD]: nice

**Comment by spxtr** [comment]
Why add `make bazel-test-integration`, if they can be run as totally normal unit tests? It seems like test size is the appropriate delimiter.

This is super awesome, btw.

**Comment by mikedanese** [comment]
@spxtr It's going to make make test run much longer is that OK do you think?

**Comment by spxtr** [comment]
You're so close!

**Comment by spxtr** [comment]
This looks super awesome to me. I'm fine with `make test` running integration tests too, but we can leave it as-is for now.

**Comment by spxtr** [comment]
/approve no-issue
@ixdy want to LGTM?

File: build/root/WORKSPACE [hunk 2]
```diff
@@ -12,6 +12,15 @@ http_archive(
     urls = ["https://github.com/kubernetes/repo-infra/archive/9dedd5f4093884c133ad5ea73695b28338b954ab.tar.gz"],
 )
 
+ETCD_VERSION = "3.0.17"
+
+new_http_archive(
```
Comment (ixdy) [inline build/root/WORKSPACE]: add the sha256 digest?

File: test/integration/framework/etcd.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+			return
+		}
+		etcdPath := filepath.Join(os.Getenv("RUNFILES_DIR"), "com_coreos_etcd/etcd")
+		// give every test the same random port each run
+		etcdPort := 20000 + rand.New(rand.NewSource(int64(adler32.Checksum([]byte(os.Args[0]))))).Intn(5000)
+		etcdURL = fmt.Sprintf("http://127.0.0.1:%d", etcdPort)
+
+		info, err := os.Stat(etcdPath)
+		if os.IsNotExist(err) {
+			glog.Fatalf("Cannot run test without etcd. Please install with ./hack/install-etcd.sh: looked in path %q", etcdPath)
```
Comment (ixdy) [inline test/integration/framework/etcd.go]: do we need this check? won't install-etcd.sh not help with bazel?

File: test/integration/framework/etcd.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	})
+}
+
+func EtcdMain(tests func() int) {
+	setupETCD()
+	os.Exit(tests())
+}
+
+// return the EtcdURL
+func GetEtcdURLFromEnv() string {
```
Comment (ixdy) [inline test/integration/framework/etcd.go]: the name of this function is somewhat incorrect now, but I guess we can cleanup later.

**Review by ixdy — COMMENTED** [review]
does this create a new etcd instance for each package?

**Comment by spxtr** [comment]
> does this create a new etcd instance for each package?

Yes, but it doesn't slow things down significantly according to Mike.

**Comment by ixdy** [comment]
I think #47140 broke the crossbuild.

**Comment by ixdy** [comment]
I opened an issue on cross-build for better visibility: #48887

**Comment by mikedanese** [comment]
@ixdy is this good to go? I addressed your comments

**Pull request merged**